flink-ml-lib/src/test/java/org/apache/flink/ml/clustering/AgglomerativeClusteringTest.java

```diff
@@ -0,0 +1,312 @@
+            Arrays.asList(
+                    Row.of(2, 5, 0, 2),
+                    Row.of(0, 4, 1.1102230E-16, 2),
+                    Row.of(3, 6, 0.0636708, 3),
+                    Row.of(1, 7, 0.1425070, 3),
+                    Row.of(8, 9, 0.3664484, 6));
+
+    private static final List<Row> MANHATTAN_AVERAGE_MERGE_INFO =
+            Arrays.asList(
+                    Row.of(0, 2, 1, 2),
+                    Row.of(3, 5, 1.5, 2),
+                    Row.of(1, 4, 3, 2),
+                    Row.of(6, 7, 3.75, 4),
+                    Row.of(8, 9, 4.875, 6));
+
+    private static final List<Row> EUCLIDEAN_SINGLE_MERGE_INFO =
+            Arrays.asList(
+                    Row.of(0, 2, 1, 2),
+                    Row.of(3, 5, 1.5, 2),
+                    Row.of(4, 7, 2.5, 3),
+                    Row.of(8, 9, 3, 6),
+                    Row.of(1, 6, 3, 3));
+
+    private static final List<Row> EUCLIDEAN_WARD_MERGE_INFO =
+            Arrays.asList(
+                    Row.of(0, 2, 1, 2),
+                    Row.of(3, 5, 1.5, 2),
+                    Row.of(1, 4, 3, 2),
+                    Row.of(6, 7, 4.2573465, 4),
+                    Row.of(8, 9, 5.5113519, 6));
+
+    private static final List<Row> EUCLIDEAN_COMPLETE_MERGE_INFO =
+            Arrays.asList(
+                    Row.of(0, 2, 1, 2),
+                    Row.of(3, 5, 1.5, 2),
+                    Row.of(1, 4, 3, 2),
+                    Row.of(6, 7, 3.3541019, 4),
+                    Row.of(8, 9, 5, 6));
+
+    private static final List<Row> EUCLIDEAN_WARD_K_AS_TWO_RESULT =
+            Arrays.asList(
+                    Row.of(Vectors.dense(1, 1), 0),
+                    Row.of(Vectors.dense(1, 4), 1),
+                    Row.of(Vectors.dense(1, 0), 0),
+                    Row.of(Vectors.dense(4, 1.5), 0),
+                    Row.of(Vectors.dense(4, 4), 1),
+                    Row.of(Vectors.dense(4, 0), 0));
+
+    private static final List<Row> EUCLIDEAN_WARD_THRESHOLD_AS_TWO_RESULT =
+            Arrays.asList(
+                    Row.of(Vectors.dense(1, 1), 0),
+                    Row.of(Vectors.dense(1, 4), 1),
+                    Row.of(Vectors.dense(1, 0), 0),
+                    Row.of(Vectors.dense(4, 1.5), 2),
+                    Row.of(Vectors.dense(4, 4), 1),
+                    Row.of(Vectors.dense(4, 0), 2));
+
+    private static final double TOLERANCE = 1e-7;
+
+    @Before
+    public void before() {
+        Configuration config = new Configuration();
+        config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true);
+        env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+        env.setParallelism(3);
+        env.enableCheckpointing(100);
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        tEnv = StreamTableEnvironment.create(env);
+        inputDataTable = tEnv.fromDataStream(env.fromCollection(INPUT_DATA)).as("features");
```

Comment:
If the input data comes from an operator whose parallelism is 3, how can we make sure that they will arrive at the `LocalAgglomerativeClusteringOperator`, whose parallelism is 1, in the same order and be assigned the same cluster ids across difference runs? If each record might be assigned a different id in different runs, why would the merge info remains the same and can be verified with an expected value?

Reply:
Good catch. I made a mistake --- and the mergeInfo should be different in different runs and only the distances should be the same accross different runs. 

I have added a `map` after `env.fromCollection()` to force it as a parallel stream and also re-wrote the corresponding unit test.